src/ServiceControl.Monitoring/Timings/TimingsStore.cs

```diff
@@ -0,0 +1,154 @@
+                {
+                    if (validIntervals.Contains(interval) == false)
+                    {
+                        MeasurementInterval bucket;
+                        endpointData.TryRemove(interval, out bucket);
+                    }
+                }
+
+            }
+        }
+
+        public EndpointTimings[] GetTimings(DateTime now)
+        {
+            var result = new List<EndpointTimings>();
+            var intervals = GenerateIntervalIds(now, NumberOfHistoricalIntervals);
+
+            foreach (var endpointName in timings.Keys)
+            {
+                ConcurrentDictionary<DateTime, MeasurementInterval> endpointData;
+
+                if (timings.TryGetValue(endpointName, out endpointData))
+                {
+                    var totalDuration = 0L;
+                    var totalMeasurements = 0L;
+                    var timingIntervals = new TimingInterval[intervals.Length];
+
+                    for (var i = 0; i < intervals.Length; i++)
+                    {
+                        timingIntervals[i] = new TimingInterval
+                        {
+                            IntervalStart = intervals[i]
+                        };
+
+                        MeasurementInterval bucket;
+
+                        if (endpointData.TryGetValue(intervals[i], out bucket))
+                        {
+                            totalDuration += bucket.TotalTime;
+                            totalMeasurements += bucket.TotalMeasurements;
+
+                            timingIntervals[i].Average = bucket.TotalTime / (double) bucket.TotalMeasurements;
+                        }
+                    }
+
+                    var endpointTimings = new EndpointTimings
+                    {
+                        EndpointName = endpointName,
+                        Average = totalMeasurements > 0 ? totalDuration / (double) totalMeasurements : 0L,
+                        Intervals = timingIntervals
+                    };
+
+                    result.Add(endpointTimings);
+                }
+            }
+
+            return result.ToArray();
+        }
+
+        static DateTime[] GenerateIntervalIds(DateTime now, int numberOfPastIntervals)
+        {
+            var intervals = new DateTime[numberOfPastIntervals];
+
+            intervals[0] = IntervalId(now);
+
+            for (var i = 1; i < numberOfPastIntervals; i++)
+            {
+                intervals[i] = intervals[i - 1].Subtract(TimeSpan.FromSeconds(IntervalSizeInSec));
+            }
+
+            return intervals;
+        }
+
+        static DateTime IntervalId(DateTime date)
+        {
+            var interval = 0;
+
+            while (interval + IntervalSizeInSec < date.Second)
+            {
+                interval += IntervalSizeInSec;
+            }
+
+            return new DateTime(date.Year, date.Month, date.Day, date.Hour, date.Minute, interval, DateTimeKind.Utc);
```

Comment:
Once again I would propose to hold-off. I think that what is missing here is a value type expressing concept of an interval. In such case the hashing algorithm would hash over intervals.

Right now the whole storage is encapsulated in a single class so I would not worry about dangerous details leaking out.